tests/rptest/tests/group_membership_test.py

```diff
@@ -113,3 +118,213 @@ def _check_no_duplicates():
+
+        check_metric()
+        metrics_offsets = self._get_offset_from_metrics(group)
+        assert metrics_offsets is None
+        self.redpanda.create_topic(topic_spec)
+        check_metric()
+
+    @cluster(num_nodes=7)
+    def test_leadership_transfer(self):
+        rpk = RpkTool(self.redpanda)
+        topics = list(
+            filter(lambda x: "MultiplePartition" in x.name, self.topics))
+        group = "g0"
+
+        producers = []
+        for topic in topics:
+            producer = RpkProducer(self._ctx,
+                                   self.redpanda,
+                                   topic.name,
+                                   msg_size=5,
+                                   msg_count=1000)
+            producer.start()
+            producers.append(producer)
+
+        consumers = []
+        for topic in topics:
+            consumer = RpkConsumer(self._ctx,
+                                   self.redpanda,
+                                   topic.name,
+                                   group=group)
+            consumer.start()
+            consumers.append(consumer)
+
+        # Wait until cluster starts producing metrics
+        wait_until(
+            lambda: self.redpanda.metrics_sample("kafka_group_offset") != None,
+            timeout_sec=10,
+            backoff_sec=5)
+
+        admin = Admin(redpanda=self.redpanda)
+
+        def get_offset_with_node_from_metric(group):
+            metric = self.redpanda.metrics_sample("kafka_group_offset")
+            if metric is None:
+                return None
+            metric = metric.label_filter(dict(group=group))
+            return metric.samples
+
+        def get_group_leader():
+            return admin.get_partitions(namespace="kafka_internal",
+                                        topic="group",
+                                        partition=0)['leader_id']
+
+        def check_metric_from_node(node):
+            metrics_offsets = get_offset_with_node_from_metric(group)
+            assert all([
+                metric.node.account.hostname == node.account.hostname
+                for metric in metrics_offsets
+            ])
+
+        def transfer_completed():
+            return get_group_leader() != -1
+
+        leader_node = self.redpanda.nodes[get_group_leader() - 1]
+        check_metric_from_node(leader_node)
+
+        # Check transfer leadership to another node
+        for i in range(3):
+            new_leader_node = random.choice(
+                list(filter(lambda x: x != leader_node, self.redpanda.nodes)))
+
+            admin.transfer_leadership_to(
+                namespace="kafka_internal",
+                topic="group",
+                partition=0,
+                target=self.redpanda.idx(new_leader_node))
+
+            wait_until(transfer_completed, timeout_sec=30, backoff_sec=5)
+
+            check_metric_from_node(new_leader_node)
+            leader_node = new_leader_node
```

Comment:
`leader_node = new_leader_node` it's error prone to assume that this is true. it's better to always query again for the leader and be prepared to handle any error case that may pop up.

this a good example of why `wait_until` is so powerful. it's best to think about your tests a set of conditions always query for updated state and are designed to eventually become true.

so I would expect most of the stuff in this overall loop to be contained inside a wait_until